### PR TITLE
feat(bootstrap_srv) iroh-relay integrated in axum routes

### PR DESCRIPTION
**Pull request by lucksus**
## Add Iroh Relay Server Integration with Rate Limiting

This PR integrates the iroh relay server into the kitsune2 bootstrap server as an axum route, enabling WebSocket-based relay functionality alongside the existing SBD signaling. 

In concert with: https://github.com/n0-computer/iroh/pull/3832

  ### Changes

  #### Core Integration
  - **Axum Route Integration**: Added `/relay` endpoint using iroh's new axum integration (through above PR against iroh) 
    - Mounts as a standard axum route alongside existing bootstrap endpoints
    - Separate router with dedicated state management
    - Added `/ping` probe endpoint for latency testing

  #### Rate Limiting
We have to do this now ourselves because axum handles connections and just forwards parsed body. Rate limiting as to be a layer of the axum route.
  - **Connection-Level Rate Limiting**: Implemented using `governor` crate with token bucket algorithm
    - Applied as tower middleware layer on relay routes
    - Configurable via `Limits` struct:
      - `accept_conn_limit`: Sustained connections per second (default: 250)
      - `accept_conn_burst`: Maximum burst capacity (default: 100)
    - Returns `429 Too Many Requests` when limits exceeded

  - **Limitations**: Byte-level rate limiting (`client_rx`) is not supported
    - Axum integration receives already-established WebSocket connections
    - No access to underlying TCP/TLS stream for byte-level control
    - Connection-level limiting provides adequate protection for this use case

  #### Dependencies
  - Added `tower = "0.5"` for middleware abstractions
  - Added `governor = "0.7"` for token bucket rate limiting
  - Updated `iroh-relay` to use forked version with axum support, and going off of 0.95.1: https://github.com/lucksus/iroh/pull/1

## Summary by CodeRabbit

* **New Features**
  * WebSocket-based /relay endpoint and lightweight /ping health probe.
  * Configurable connection rate limiting with sensible defaults, validation, and 429 responses on exceedance.

* **Chores**
  * Replaced legacy reverse-proxy wiring with an in-process Axum-based relay integration and optional WebSocket transport.
  * Updated and pinned relay-related dependencies for testing and dev.

* **Tests**
  * Added tests covering relay flows and rate-limiter edge cases.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -160,3 +162,5 @@ jsonschema = "0.30.0"
 #sbd-e2e-crypto-client = { path = "../sbd/rust/sbd-e2e-crypto-client" }
 #tx5-core = { path = "../tx5/crates/tx5-core" }
 #tx5 = { path = "../tx5/crates/tx5" }
+iroh = { git = "https://github.com/lucksus/iroh.git", rev = "d15e0104fd7e17f2dba15e9ff41ed8e7cf30043b" }
+iroh-relay = { git = "https://github.com/lucksus/iroh.git", rev = "d15e0104fd7e17f2dba15e9ff41ed8e7cf30043b" }
```
Comment (lucksus) [inline Cargo.toml]: This PR is at above rev and has all the same changes as the one against n0-computer but is going off of 0.95.1 so everything still builds with that latest iroh release: https://github.com/lucksus/iroh/pull/1

File: crates/bootstrap_srv/src/iroh_relay_axum.rs [hunk 2]
```diff
@@ -0,0 +1,796 @@
...
+    State(state): State<RelayState>,
+    ws: WebSocketUpgrade,
+    headers: HeaderMap,
+) -> Response {
+    // Extract the client auth header if present
+    let client_auth_header = headers
+        .get(iroh_relay_holochain::http::CLIENT_AUTH_HEADER)
+        .cloned();
+
+    debug!("Relay WebSocket upgrade request");
```
Comment (ThetaSinner) [inline crates/bootstrap_srv/src/iroh_relay_axum.rs]: Log messages in this function should attempt to identify the client - anything on `ws` or headers that might be relevant?

File: crates/bootstrap_srv/src/iroh_relay_axum.rs [hunk 3]
```diff
@@ -0,0 +1,796 @@
...
+            Poll::Ready(Err(e)) => Poll::Ready(Err(StreamError::Io(
+                std::io::Error::other(format!("{:?}", e)),
+            ))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+// Axum WebSocket doesn't support TLS key export, so we return None
+impl ExportKeyingMaterial for AxumWebSocketAdapter {
```
Comment (ThetaSinner) [inline crates/bootstrap_srv/src/iroh_relay_axum.rs]: I'll keep looking with this one but this strikes me as odd. This is an internal trait, so why have it if it's not implemented here?

File: crates/transport_iroh/src/connection_context.rs [hunk 4]
```diff
@@ -158,27 +157,22 @@ impl ConnectionContext {
...
-            }
-            None => ConnectionType::None,
-        }
+    /// Check if the connection has any direct (IP-based) path.
+    /// Returns true if at least one path is using a direct IP address.
+    pub fn is_direct(&self) -> bool {
+        // For now, we can't easily check the paths without async access to the connection.
+        // We'll return false as a conservative default.
+        // This could be improved by storing path information when the connection is created.
+        false
```
Comment (mattyg) [inline crates/transport_iroh/src/connection_context.rs]: Seems like this doesn't work anymore?

File: crates/bootstrap_srv/src/iroh_relay_axum.rs [hunk 5]
```diff
@@ -0,0 +1,796 @@
+//! Axum integration for the iroh relay server.
+//!
+//! This module provides an axum-compatible handler that can be mounted
+//! as a standard route, avoiding the need for connection-level routing.
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/bootstrap_srv/src/iroh_relay_axum.rs]: That's kinda the point of this whole PR and the PR I had put to iroh: their code uses hyper and thus as access to the connection itself, which they use to level up an http request to a websocket, with TLS.

Axum, being a crate for servers, does that internally and hides it away for its user. You just write a handler for a route.

This crate here (which needed the exports of some private types in iroh) provides a way to put iroh into axum without loosing the ability to upgrade the connection...

The alternative would have been "connection-level routing" meaning, to do hyper stuff before the connection reaches axum and check if the rout is `/relay` and pipe it into iroh.

File: crates/transport_iroh/src/test_utils/harness.rs [hunk 6]
```diff
@@ -25,20 +25,25 @@ impl IrohTransportTestHarness {
         }
         .with_default_config()
         .unwrap();
-        let (_, relay_url, relay_server) = spawn_iroh_relay_server().await;
+
+        // Spawn a test bootstrap server with integrated relay support
+        let bootstrap_server = TestBootstrapSrv::new(false).await;
+        // Note: relay_url must end with '/' for proper URL construction
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/transport_iroh/src/test_utils/harness.rs]: https://github.com/holochain/kitsune2/pull/432/changes/c10ce1a03b79eb14c5f48ef5a727c66219f0c7eb

File: crates/transport_iroh/src/connection_context.rs (hunk 4 above)
Comment (lucksus) [reply crates/transport_iroh/src/connection_context.rs]: https://github.com/holochain/kitsune2/pull/432/commits/8079879e29910551eb69809916e76be66ff90f4a

File: crates/transport_iroh/src/connection.rs [hunk 7]
```diff
@@ -67,6 +70,12 @@ impl Connection for IrohConnection {
     fn close(&self, code: u8, reason: &[u8]) {
         self.inner.close(code.into(), reason);
     }
+
+    fn is_direct(&self) -> bool {
```
Comment (mattyg) [inline crates/transport_iroh/src/connection.rs]: Is there a way to test this?

File: crates/transport_iroh/src/lib.rs [hunk 8]
```diff
@@ -359,6 +356,20 @@ impl IrohTransport {
         config: IrohTransportConfig,
         handler: Arc<TxImpHnd>,
     ) -> K2Result<DynTxImp> {
+        // Ensure the relay URL ends with '/' so that iroh appends
+        // paths correctly rather than replacing the last segment.
+        // Normalize early so all downstream uses see the same URL.
+        let config = IrohTransportConfig {
```
Comment (jost-s) [inline crates/transport_iroh/src/lib.rs]: I think this is better done when validating the config.

**Review by jost-s — CHANGES_REQUESTED** [review]
The relay integration looks good to me.

There are several unrelated changes which I request to remove or otherwise argue for their benefit.

File: crates/transport_iroh/src/lib.rs [hunk 9]
```diff
@@ -618,8 +624,21 @@ impl TxImp for IrohTransport {
         let connections = self.connections.clone();
         let connection_locks = self.connection_locks.clone();
 
+        // Parse the configured relay URL if present
+        let configured_relay = self.config.relay_url.as_ref().and_then(|url| {
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/transport_iroh/src/lib.rs]: The problem is that the relay is now under a path (and not just a domain name + port).

`endpoint_from_url` takes a kitsune2 peer URL like `http://127.0.0.1:12345/<endpoint_id>` and must reconstruct the iroh EndpointAddr which includes the RelayUrl. The old code reconstructed it as http://127.0.0.1:12345/ — but now that the relay is served at a path `/relay`.

When iroh sees http://127.0.0.1:12345/ (without /relay/ - what the old code did) as the relay URL, it doesn't recognize it as the same relay and creates a second relay actor.

File: crates/transport_iroh/src/connection_context.rs [hunk 10]
```diff
@@ -158,26 +155,21 @@ impl ConnectionContext {
...
-            abort_handle.abort();
+        match self.connection_reader_abort_handle.try_lock() {
+            Ok(mut guard) => {
+                if let Some(abort_handle) = guard.take() {
+                    abort_handle.abort();
+                }
+            }
+            Err(_) => {
+                warn!("Could not acquire lock to abort connection reader task");
+            }
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/transport_iroh/src/connection_context.rs]: remnant from a different approach. reverted.

File: crates/transport_iroh/src/connection_context.rs [hunk 11]
```diff
@@ -23,7 +21,7 @@ use tracing::{debug, error, info, trace, warn};
 pub(super) struct ConnectionContext {
     handler: Arc<TxImpHnd>,
     connection: DynConnection,
-    connection_reader_abort_handle: Mutex<Option<AbortHandle>>,
+    connection_reader_abort_handle: Arc<Mutex<Option<AbortHandle>>>,
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/transport_iroh/src/connection_context.rs]: remnant from a different approach. reverted.

File: crates/transport_iroh/src/endpoint.rs [hunk 12]
```diff
@@ -1,26 +1,29 @@
...
 impl<W> EndpointAddrWatcher for IrohWatcher<W>
 where
-    W: Watcher<Value = EndpointAddr> + Send + Sync,
+    W: n0_watcher::Watcher<Value = EndpointAddr> + Send + Sync + Unpin,
 {
     fn updated(&mut self) -> BoxFut<'_, Result<EndpointAddr, Disconnected>> {
-        Box::pin(self.inner.updated())
+        Box::pin(async { n0_watcher::Watcher::updated(&mut self.inner).await })
     }
 }
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/transport_iroh/src/endpoint.rs]: reverted

File: crates/bootstrap_srv/src/lib.rs [hunk 13]
```diff
@@ -221,7 +221,10 @@ pub use auth::*;
 mod sbd;
 
 #[cfg(feature = "iroh-relay")]
-mod iroh_relay;
+pub mod iroh_relay;
```
[the remark this answers is not in the document]
Comment (lucksus) [reply crates/bootstrap_srv/src/lib.rs]: I'm still not sure I understand what you're saying. Taking out this mention of the sub-module breaks the build, because it's used - both in this crate, and also (why we need the pub) in other crates.

```
   Compiling kitsune2_bootstrap_srv v0.4.0-dev.2 (/home/lucksus/Coasys/code/kitsune2/crates/bootstrap_srv)
error[E0433]: failed to resolve: could not find `iroh_relay` in the crate root
   --> crates/bootstrap_srv/src/http.rs:265:46
    |
265 |                     let relay_state = crate::iroh_relay::create_relay_state();
    |                                              ^^^^^^^^^^ could not find `iroh_relay` in the crate root

error[E0433]: failed to resolve: could not find `iroh_relay` in the crate root
   --> crates/bootstrap_srv/src/http.rs:267:62
    |
267 |                         .route("/relay", routing::get(crate::iroh_relay::relay_handler))
    |                                                              ^^^^^^^^^^ could not find `iroh_relay` in the crate root
```

without `pub`:
```
Compiling kitsune2_test_utils v0.4.0-dev.2 (/home/lucksus/Coasys/code/kitsune2/crates/test_utils)
error[E0603]: module `iroh_relay` is private
   --> crates/test_utils/src/bootstrap.rs:17:50
    |
 17 |     _relay_state: Option<kitsune2_bootstrap_srv::iroh_relay::RelayState>,
    |                                                  ^^^^^^^^^^  ---------- struct `RelayState` is not publicly re-exported
    |                                                  |
    |                                                  private module
```
